### ISSUE TITLE
There are a lot of things that seems to be different about opencv-rust vs the opencv C++

### ISSUE THREAD

**Issue by Joe23232** [twistedfall/opencv-rust#211]
Not trying to come off as rude or anything but according to this [C++ OpenCV video](https://youtu.be/2FYm3GOonhk) there are a lot of things that are just different about `opencv-rust` compared to opencv C++.

For example in this [part of the video](https://youtu.be/2FYm3GOonhk?t=3728), in order to produce an only blue screen the guy types this out:

```c++
Mat img(512, 512, CV_8UC3, Scalar(255, 0, 0);
```

I tried to search it in the [Rust docs](https://docs.rs/opencv/0.49.1/opencv/core/struct.Mat.html), however this example is for C++ only, and I am not too sure what is the Rust's equivalent.

Am I missing something?

**Comment by twistedfall** [comment]
The equivalent for this particular line would be:
```rust
let img = Mat::new_rows_cols_with_default(512, 512, Vec3b::typ(), Scalar::new(255, 0, 0, 0));
```

**Comment by Joe23232** [comment]
Ah I see, Thanks :), but just curious wouldn't it make harder for people to migrate from C++ to Rust for `opencv` if the syntax is different?

Edit: Is there a document page which is similar to the [Rust's book](https://doc.rust-lang.org/book/) and [Bevy Game Engine's book](https://bevyengine.org/learn/book/introduction/)?

**Comment by twistedfall** [comment]
> wouldn't it make harder for people to migrate from C++ to Rust for opencv if the syntax is different?

Can you give me an example of how e.g. this particular example can be implemented better? I'm always open to the improvement suggestions!

> Edit: Is there a document page which is similar to the Rust's book and Bevy Game Engine's book?

Not really, I'm not sure I have resources to create something like this, help is always welcome! The only crate-specific documentation at the moment is the [readme file](https://github.com/twistedfall/opencv-rust/blob/master/README.md).

**Comment by Joe23232** [comment]
@twistedfall 

> Can you give me an example of how e.g. this particular example can be implemented better? I'm always open to the improvement suggestions!

While unlike majority of Rust's crates out there where they have messy implementations, I like how this is not too difficult to use.

However as mentioned earlier this crate compared to Python and C++'s API is a bit too different in a lot of areas where it makes it very hard to follow along a Python or/and a C++ tutorial using OpenCV. Even Python and C++'s APIs are very similar to each other compared to `opencv-rust`.

Here are some of the examples in which is quite diffierent:

1. When trying to use a specific colour for a background, in C++ implementation it is this:

```c++
Mat img(512, 512, CV_8UC3, Scalar(255, 0, 0);
```

Whereas in Rusts implementation it is this:

```rust
let img = Mat::new_rows_cols_with_default(512, 512, Vec3b::typ(), Scalar::new(255, 0, 0, 0));
```

Instead to make it similar, wouldn't it be better if it was something like this?

```rust
let blue_background = Mat::img(512, 512, CV_8UC3, Scalar(255, 0, 0));
```

So like `img` is the function instead of `new_rows_cols_with_default()`. Just curious why is it named completely different?

Another example in regards to importing videos:

```c++
string path = "video.jpg";
VideoCapture cap(path);
```
Rust,s example:

```rust
pub fn from_file(filename: &str, api_preference: i32) -> Result<VideoCapture> // This is the format from the docs
```

While yes the C++ version is a little confusing as to what they are doing compared to Rust, however since the names are completely different, it makes it harder to learn the Rust's naming conventions, maybe this could be a good suggestion if it was done like this:

```rust
let cap = imgcodecs::video_capture("video.jpg").unwrap();
```

Something similar to make it easy to port over from C++ to Rust.

Also as a suggestion I am not too sure if this is possible or not but in this example:

```rust
let image = imgcodecs::imread(r"opencv-course\Resources\Photos\cat.jpg", IMREAD_COLOR).expect("Cannot file image or image is corrupted!");
    
highgui::named_window("hello opencv!", WINDOW_AUTOSIZE).unwrap();
highgui::imshow("hello opencv!", &image).unwrap();
```

Isn't it possible to get rid of the `highgui` (and other parent names) like in Python and C++? I know its not a huge deal and this is something I am ok with but it could make it a little nicer and more simple? So we then instead end up seeing:

```rust
named_window("hello opencv!", WINDOW_AUTOSIZE).unwrap();
imshow("hello opencv!", &image).unwrap();
```

In regards to showing an image on the screen, in C++ vs Rust:

```c++
Mat img = imread("image.jpg");
imshow("Image", img)
waitKey(0);
```

```rust
let image = imgcodecs::imread(r"image.jpg", IMREAD_COLOR).expect("Cannot file image or image is corrupted!");
    
// Why is there an extra function I have to call just to display it?
highgui::named_window("hello opencv!", WINDOW_AUTOSIZE).unwrap();
highgui::imshow("hello opencv!", &image).unwrap();

highgui::wait_key(10000).unwrap();
```

Why do I see two extra functions? Why not just have one function which is:
```rust
highgui::imshow("hello opencv!", &image, WINDOW_AUTOSIZE).unwrap();
```


Another example is in regards to waiting

```c++
waitKey(0);
```

```rust
highgui::wait_key(10000).unwrap();
```

While this is not such a huge deal, how come over here it is `10000` but for C++ and Python it is `0` or am I missing something?

I apologise if I sounded too harsh or anything, my intention is to provide only suggestions from my point of view. I am sure you know much more than I do as I am a beginner but yeah I hope this helps :)

---

> Not really, I'm not sure I have resources to create something like this, help is always welcome!

All good I understand :)

**Comment by twistedfall** [comment]
> Even Python and C++'s APIs are very similar to each other compared to opencv-rust.

It's debatable of course, but I don't think it's exactly true. For example in Python bindings there is no `Mat` at all, you need to use `ndarray` instead.

> So like img is the function instead of new_rows_cols_with_default(). Just curious why is it named completely different?

In this example the `img` is the variable name, not a function. The function called is a constructor and they are conventinally named `new` in Rust. Since there is no function overloading (without tricks) in Rust, several overloaded constructors from C++ need to be translated into methods with different names.

> While yes the C++ version is a little confusing as to what they are doing compared to Rust, however since the names are completely different, it makes it harder to learn the Rust's naming conventions, maybe this could be a good suggestion if it was done like this:

It's again about the constructor, `VideoCapture` has several constructors, I don't really see how having a function that would create a `VideoCapture` instance would solve the need to be able to access all of its constructors.

> Isn't it possible to get rid of the highgui (and other parent names) like in Python and C++?

For that you just need to import the function itself, not the module like ```use opencv::highgui::imshow```.

> // Why is there an extra function I have to call just to display it?

You don't have too, you can drop it. The crate functions are just automatically generated wrappers for the corresponding OpenCV functions with proper type marshaling.

> Why do I see two extra functions?

I can only see one, a call to `named_window` which one is the second?

> Why not just have one function which is:

Short answer, because such function (with those 3 arguments) doesn't exist in OpenCV. To keep the support burden down I try to avoid manually implementing functions that extend the functionality of OpenCV, I'd like to keep it strictly 1-to-1 function mapping as far as possible. There are some manual implementations, but those are limited to areas which are not (yet) possible to translate automatically. E.g. `Mat_<T>` struct.

> While this is not such a huge deal, how come over here it is 10000 but for C++ and Python it is 0?

It's not, you're getting different result by calling this function with 10000 against 0. Check the docs about the meaning of that argument: https://docs.rs/opencv/0.49.1/opencv/highgui/fn.wait_key.html

**Comment by Joe23232** [comment]
> It's debatable of course, but I don't think it's exactly true. For example in Python bindings there is no Mat at all, you need to use ndarray instead.

Interesting

> In this example the img is the variable name, not a function. The function called is a constructor and they are conventinally named new in Rust. Since there is no function overloading (without tricks) in Rust, several overloaded constructors from C++ need to be translated into methods with different names.

Oh that makes sense now.

> For that you just need to import the function itself, not the module like use opencv::highgui::imshow.

How would I do this?

> Short answer, because such function (with those 3 arguments) doesn't exist in OpenCV. To keep the support burden down I try to avoid manually implementing functions that extend the functionality of OpenCV, I'd like to keep it strictly 1-to-1 function mapping as far as possible. There are some manual implementations, but those are limited to areas which are not (yet) possible to translate automatically. E.g. Mat_<T> struct.

I see.

> It's not, you're getting different result by calling this function with 10000 against 0. Check the docs about the meaning of that argument: https://docs.rs/opencv/0.49.1/opencv/highgui/fn.wait_key.html

Oh right I get it now mate.

---

WHat if the docs showed Rusts alternative to C++ APIs? For with stuff like this:

```c++
Mat img(512, 512, CV_8UC3, Scalar(255, 0, 0);
```
the Rust's version would be

```rust
let img = Mat::new_rows_cols_with_default(512, 512, Vec3b::typ(), Scalar::new(255, 0, 0, 0));
```

?

**Comment by twistedfall** [comment]
> How would I do this?

I've written it in my reply: `use opencv::highgui::imshow`

> WHat if the docs showed Rusts alternative to C++ APIs? 

That's a huge amount of manual work and I don't really see how this is going to help in discoverability. Usually you want to translate C++ to Rust, but such annotation will be mostly useful in translating Rust to C++.

**Comment by Joe23232** [comment]
@twistedfall 

>I've written it in my reply: use opencv::highgui::imshow

Why not just do this instead?
```rust
use opencv::highgui::*;
use opencv::imgcodecs::*;
```

This seems to work fine for me.

> That's a huge amount of manual work and I don't really see how this is going to help in discoverability. Usually you want to translate C++ to Rust, but such annotation will be mostly useful in translating Rust to C++.

I see understandable.

**Comment by twistedfall** [comment]
This of course is also an option as long as there are no conflicting items.
